library/authenticate/ftrack/library/authenticate/authenticate.py

```diff
@@ -0,0 +1,149 @@
+# :coding: utf-8
+# :copyright: Copyright (c) 2024 ftrack
+
+import logging
+import webbrowser
+import threading
+from typing import TYPE_CHECKING, Optional
+from .util.identifier import generate_url_identifier
+from ftrack.library.utility.url.checker import (
+    url_checker,
+    ftrack_server_url_checker,
+)
+
+if TYPE_CHECKING:
+    from .helper.credential import (
+        CredentialProviderFactory,
+        CredentialProvider,
+    )
+    from .helper.webserver import WebServerFactory, WebServer
+
+# Configure logging
+logging.basicConfig(level=logging.INFO)
+
+
+class Authenticate:
+    """
+    Handles authentication by opening a browser and running a local web server
+    to capture credential via callback.
+    """
+
+    def __init__(
+        self,
+        server_url: str,
+        credential_provider_factory: "CredentialProviderFactory",
+        web_server_factory: "WebServerFactory",
+        redirect_port: int = 5000,
+    ) -> None:
+        """
+        Initialize the Authenticate instance.
+
+        :param server_url: The base URL for the authentication request.
+        :param credential_provider_factory: The credential provider factory instance.
+        :param web_server_factory: The web server factory instance.
+        :param redirect_port: The port on which the local web server will run.
+        """
+        try:
+            self._server_url: str = url_checker(
+                server_url, [ftrack_server_url_checker]
+            )
+        except ValueError as e:
+            logging.error(f"Invalid server URL: {e}")
+            raise
+        self._credential_provider_instance: "CredentialProvider" = (
+            credential_provider_factory.create_credential_provider()
+        )
+        self._web_server_factory: "WebServerFactory" = web_server_factory
+        self._web_server_instance: Optional["WebServer"] = None
+        self._redirect_port: int = redirect_port
+        self._redirect_uri: str = f"http://localhost:{redirect_port}/callback"
+        self._server_ready: threading.Event = threading.Event()
+
+    @property
+    def server_url(self):
+        return self._server_url
+
+    @property
+    def credential_provider_instance(self):
+        return self._credential_provider_instance
+
+    @property
+    def web_server_factory(self):
+        return self._web_server_factory
+
+    @property
+    def web_server_instance(self):
+        return self._web_server_instance
+
+    @property
+    def redirect_port(self):
+        return self._redirect_port
+
+    @property
+    def redirect_uri(self):
+        return self._redirect_uri
+
+    def browser_authenticate(self) -> None:
```

Comment:
I don't see the difference?
I mean I currently have browser_authenticate and credential_authenticate as methods. why flipping the order will make it best practice?

Reply:
Agreed to change this to authenticate_browser and authenticate_credential. Because this way you can .authenticate_... and it will show you the available possibilities.